### ISSUE TITLE
Discussion:  how to handle upgrades

### ISSUE THREAD

**Issue by roberts-pumpurs** [equilibriumco/axelar-amplifier-solana#17]
**Summary**
A private key strategy should be finalized and ideally offline signing mechanisms, CI/CD pipelines should be considered, for robust and secure processes around deploying/updating the contracts on testnet/mainnet.

**Comment by eloylp** [comment]
I would like to add a bit more of information to this issue, originated from this [slack](https://eig3r.slack.com/archives/C06A2QAPE2Z/p1722341271621779?thread_ts=1722324504.466299&cid=C06A2QAPE2Z) thread discussion.

Axelar is expecting that all the contracts (Gateway, ITS) addresses are renewed on each update. But client dapps should have a stable reachable address. In the EVM world, this is achieved by "proxy contracts", which ultimately forward the smart contract call to the right, updated contract deployment address.

We don't actually know of a similar mechanism in the Solana world. So we need to investigate alternatives that allow us to have the same address, but different smart contract deployments under the hood.

As commented in slack thread discussion, it's probably very recommended to reach them regarding this and re-asses they maintain this invariant for the contracts and that we are in the right path investigating this.

**Comment by eloylp** [comment]
A possible good read: https://solana.com/developers/evm-to-svm/complete-guide

**Comment by eloylp** [comment]
Let's also take into account that for generating Solana PDA's , we are [adding the address program as part of the init vector](https://github.com/eigerco/solana-axelar-internal/blob/f43210712a250c71708af0c0322fc99f94eb9299/solana/programs/gateway/src/processor/initialize_config.rs#L39-L40). That means any change in programs address would result in diffrent PDA's, thus different state.

At this point changing the address on each deployment doesn't seem a doable option.

**Comment by eloylp** [comment]
When upgrading contracts with the Solana CLI, a proxy strategy is already implemented, so a similar process like eth is happening, thus , [this](https://github.com/eigerco/solana-axelar/issues/17) should not be a problem anymore.

**Comment by roberts-pumpurs** [comment]
> When upgrading contracts with the Solana CLI, a proxy strategy is already implemented, so a similar process like eth is happening

Yes, this is correct. This is a good resource on what the [solana CLI does](https://solana.com/docs/programs/deploying#how-object-object-works). And we have a test in our code that replicates all the steps (if you prefer reading Rust over docs 😄 ). And once we develop the governance contract, it will be responsible for doing contract upgrades eigerco/solana-axelar-internal#278 eigerco/solana-axelar-internal#280 so we don't have to worry about the private keys post-deployment.


> That means any change in programs address would result in diffrent PDA's, thus different state.

Yes, that is also correct. My understanding is that, once deployed, our Solana Gateway address should never change.

**Comment by frenzox** [comment]
A reminder: Whenever we have some sort of automated deployment in place (or even non-automated), we need to make it trigger the e2e tests.

**Comment by eloylp** [comment]
Maybe a CI/CD strategy is not possible for testnet. But in order to better explore the needs we could follow @frenzox recommendation of using https://github.com/release-plz/release-plz . We could use it from a terminal, see how it works manually for the first depoyments and then automate. WDYT ? 

@frenzox @ICavlek

**Comment by eloylp** [comment]
I am also, having this doubt: Since the bindings are going to be the main interface, how we are going to version, pack and publish them ? @frenzox @ICavlek

**Comment by ICavlek** [comment]
> I am also, having this doubt: Since the bindings are going to be the main interface, how we are going to version, pack and publish them

The way I see it, is when an update in solana programs API is happening, bindings also need to be updated. Therefore, I don't see any difference from other things, meaning that bindings could also go under `release-plz`.

It could be argued about bindings being a separate product, but I wouldn't overengineer things here and leave it as it is.

**Comment by eloylp** [comment]
> The way I see it, is when an update in solana programs API is happening, bindings also need to be updated. Therefore, I don't see any difference from other things, meaning that bindings could also go under release-plz.

I might be wrong, but i think `release-plz` is only for releasing Rust crates. The use case i have in mind for the bindings  is:

> Me as an user, i want to import the bindings easily on my JS project by just 'npm install @eiger/bindings' , being ready to go for doing GMP calls.

Let me know if am missing something or assuming too many things ™

**Comment by ICavlek** [comment]
> Me as an user, i want to import the bindings easily on my JS project by just 'npm install @eiger/bindings' , being ready to go for doing GMP calls.

@eloylp This is a valid point, you are not assuming too many things or missing anything.
Bindings and the better-experience-UI-UX-wrappers around them are already prepared TS/JS code, which means that it should be doable to publish them as you have suggested.

> npm install @eiger/bindings

I am completely fine with this setup, maybe `@eiger/solana-axelar-bindings.`
Additionally,  due to the possible changes in solana programs without touching the interface, this means that at some point bindings and the rest of the crates are going to diverge in version numbering at some point.

**Comment by eloylp** [comment]
Ok, so we will need an npm registry account for publishing things and another one for crates.io 📝

**Comment by frenzox** [comment]
I stumbled into [`cargo-dist`](https://github.com/axodotdev/cargo-dist) while looking for a way to package the deployment scripts, might be worth looking into.

**Comment by eloylp** [comment]
Writing here an update about the state of the art automated deployments roadmap:
1. We are going to use https://github.com/release-plz/release-plz for publishing the rust contracts into crates.io. That way rust instruction builders should be reachable.
2. Regarding the bindings, currently the only way of importing them is to clone the repo and [add it this way in the package.json](https://github.com/eigerco/axelar-solana-e2e/blob/main/package.json#L7). But for mainnet, we would like to have an official npm published package for them. We need to maintain a correlation among bindings versions and rust versions - probably always publishing npm packages no matter if there are changes in public api, so we keep things simple for the final user.
3. In regards `1` we would like to automate the process, but we have doubts about how the CI would work with the multisig. Probably it should be done in 2 steps. The first one (CI) would just leave prepared the multisig transaction ready to be signed. After all member sign, any of them can do the final triggering of the update. But i am just thinking loud here.
4. Due to `1` and `3` our philosophy right now is to act on demand. Maybe do first manual releases to better identify the parts that should be automated.    

Let me know if am missing something

**Comment by eloylp** [comment]
We have created https://github.com/eigerco/solana-axelar/issues/147 for addressing the bindings npm publishing part before testnet deployment, as the users will be using the JS api at most.

Before `mainnet` we should also pusblish the RUST api as well at https://github.com/eigerco/solana-axelar/issues/148

(consensuated after internal dicsussions)
